File: legacy/iRecHS2/scripts/irechs2.m
classdef irechs2 < handle
    % iRecHS2 TCP client for the iRecHS2 eye tracker

    properties (Access = private)
        HOST
        PORT
        State = "disconnect"
        RemainStr = ""
        Client = []
    end

    methods
        function obj = irechs2(host, port)
            if nargin < 1 || isempty(host)
                host = "192.168.1.50";
            end
            if nargin < 2 || isempty(port)
                port = 35358;
            end
            obj.HOST = host;
            obj.PORT = port;
            obj.connect();
        end

        function s = state(obj)
            s = char(obj.State);
        end

        function connect(obj)
            if obj.State ~= "disconnect"
                return;
            end
            obj.Client = [];
            obj.State = "connect";
            try
                obj.Client = tcpclient(obj.HOST, obj.PORT, "Timeout", 5);
            catch ME
                if ~isempty(obj.Client)
                    try
                        clear obj.Client;
                    catch
                    end
                end
                obj.State = "disconnect";
                disp("timeout 5sec");
                disp(ME.message);
                disp(obj.HOST + ":" + string(obj.PORT));
            end
        end

        function send(obj, h, v, s, cl)
            if obj.State ~= "connect" && obj.State ~= "receive"
                return;
            end
            gp = sprintf("calibration\n%10.4f,%6.2f,%6.2f,%6.2f\n", cl, h, v, s);
            write(obj.Client, uint8(gp), "uint8");
        end

        function start(obj)
            if obj.State ~= "connect"
                return;
            end
            obj.State = "receive";
            write(obj.Client, uint8("start"), "uint8");
        end

        function start_plus(obj)
            if obj.State ~= "connect"
                return;
            end
            obj.State = "receive";
            write(obj.Client, uint8("start+"), "uint8");
        end

        function stop(obj, forceFlush)
            if nargin < 2
                forceFlush = true;
            end
            if obj.State ~= "receive"
                return;
            end
            obj.RemainStr = "";
            write(obj.Client, uint8("stop"), "uint8");
            if forceFlush
                pause(0.1);
                obj.get();
            end
            obj.State = "connect";
        end

        function close(obj)
            obj.stop();
            if obj.State == "connect"
                pause(0.1);
                if ~isempty(obj.Client)
                    clear obj.Client;
                    obj.Client = [];
                end
                obj.State = "disconnect";
            end
            obj.RemainStr = "";
        end

        function df = get(obj)
            df = table([], [], [], [], [], ...
                "VariableNames", {"time", "h", "v", "s", "openness"});
            if obj.State ~= "receive"
                return;
            end
            try
                if obj.Client.NumBytesAvailable == 0
                    return;
                end
                data = read(obj.Client, obj.Client.NumBytesAvailable, "uint8");
            catch
                obj.State = "disconnect";
                if ~isempty(obj.Client)
                    clear obj.Client;
                    obj.Client = [];
                end
                return;
            end

            if isempty(data)
                obj.State = "disconnect";
                if ~isempty(obj.Client)
                    clear obj.Client;
                    obj.Client = [];
                end
                return;
            end

            s = obj.RemainStr + string(char(data(:)'));
            parts = split(s, newline);
            if ~endsWith(s, newline)
                obj.RemainStr = parts(end);
                parts(end) = [];
            else
                obj.RemainStr = "";
            end

            if isempty(parts)
                return;
            end

            rows = [];
            for i = 1:numel(parts)
                line = parts(i);
                if count(line, ",") ~= 4
                    continue;
                end
                vals = sscanf(line, "%f,%f,%f,%f,%f");
                if numel(vals) == 5
                    rows = [rows; vals']; %#ok<AGROW>
                end
            end

            if ~isempty(rows)
                df = array2table(rows, "VariableNames", ...
                    {"time", "h", "v", "s", "openness"});
            end
        end
    end
end

File: legacy/iRecHS2/scripts/irechs2.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-
"""
iRecHS2 python class module

"""
from __future__ import division
from __future__ import unicode_literals

import pandas as pd
from io import StringIO
import numpy as np
import socket
import time


class iRecHS2(object):

    def state(self):
        return self.__state

    def connect(self):
        if self.__state=='disconnect':
            self.__client_socket = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
            self.__client_socket.settimeout(5)
            self.__state='connect'
            try:
                self.__client_socket.connect((self.HOST,self.PORT))
            except OSError as msg:
                timeout=self.__client_socket.gettimeout()
                print('timeout {}sec'.format(timeout))
                print(msg)
                print('{}:{}'.format(self.HOST,self.PORT))
                self.__client_socket.close()
                self.__state='disconnect'

    # def __init__(self,host='localhost',port=35358):
    def __init__(self,host='192.168.1.50',port=35358):
        self.HOST=host
        self.PORT=port
        self.__remainStr=''
        self.__state='disconnect'
        self.connect()

    def send(self,h,v,s,cl):
        if (self.__state=='connect') or (self.__state=='receive'):
            gp='calibration\n'+'{t:>10.4f},{h:>6.2f},{v:>6.2f},{s:>6.2f}\n'.format(t=cl,h=h,v=v,s=s)
            self.__client_socket.sendall(bytes(gp,"ascii"))
    def start(self):
        if (self.__state=='connect'):
            self.__state='receive'
            self.__client_socket.sendall(b'start')

    def start_plus(self):
        if (self.__state=='connect'):
            self.__state='receive'
            self.__client_socket.sendall(b'start+')

    def stop(self,forceFlush=True):
        if (self.__state=='receive'):
            self.__remainStr=''
            self.__client_socket.sendall(b'stop')
            if forceFlush:
                time.sleep(0.1)
                self.get()     
            self.__state='connect'

    def close(self):
        self.stop()
        if (self.__state=='connect'):
            time.sleep(0.1)
            self.__client_socket.close()
            self.__state='disconnect'
        self.__remainStr=''

    def get(self):
        df=pd.DataFrame([],columns=['time','h','v','s','openness'])
        if (self.__state!='receive'):
            return df
        try:
            data=self.__client_socket.recv(4096)
        except OSError:
            self.__state='disconnect'
            self.__client_socket.close()
            return df
        if data==b'':
            self.__state='disconnect'
            self.__client_socket.close()
            return df
        s=self.__remainStr+data.decode()
        sp=s.split('\n')
        if (s[-1]!='\n'):
            self.__remainStr=sp.pop(-1)
        else:
            self.__remainStr=''
        for line in sp:        
            if (line.count(',')==4):
                df=df.append(pd.read_csv(StringIO(line),names=['time','h','v','s','openness'],dtype=float),ignore_index=True)
            else:
                continue
        return df

if __name__ == '__main__':

    # eye=iRecHS2('127.0.0.1') 
    eye=iRecHS2('192.168.1.50') 
    if eye.state()=='connect':
        print('---start---')
        eye.start()
        df=eye.get()
        print(df)
        print('---pause 5[sec]--')
        eye.stop()
        time.sleep(5)
        print('---data after start--')        
        eye.start()
        df=eye.get() 
        print(df)
        print('---close---')
        eye.close()
        print('---reconnect---')
        eye.connect()
        eye.start()
        df=eye.get()
        print(df)
        print('---close---')
        eye.close()
